Cython/Utility/StringTools.c

```diff
@@ -665,31 +693,67 @@ static int __Pyx_PyBytes_Tailmatch(PyObject* self, PyObject* substr,
 
 static int __Pyx_PyBytes_SingleTailmatch(PyObject* self, PyObject* arg,
                                          Py_ssize_t start, Py_ssize_t end, int direction) {
-    const char* self_ptr = PyBytes_AS_STRING(self);
-    Py_ssize_t self_len = PyBytes_GET_SIZE(self);
-    const char* sub_ptr;
+    char* self_ptr;
+    Py_ssize_t self_len;
+    char* sub_ptr;
     Py_ssize_t sub_len;
     int retval;
 
+    #if CYTHON_COMPILING_IN_LIMITED_API && __PYX_LIMITED_VERSION_HEX < 0x030B0000
+    PyObject *converted_arg = NULL;
+    #else
     Py_buffer view;
     view.obj = NULL;
+    #endif
 
-    #if !CYTHON_ASSUME_SAFE_SIZE
-    if (unlikely(self_len < 0)) return -1;
+    #if !(CYTHON_ASSUME_SAFE_MACROS && CYTHON_ASSUME_SAFE_SIZE)
+    if (PyBytes_AsStringAndSize(self, &self_ptr, &self_len) == -1) return -1;
+    #else
+    self_ptr = PyBytes_AS_STRING(self);
+    self_len = PyBytes_GET_SIZE(self);
     #endif
 
     if (PyBytes_Check(arg)) {
+        #if !(CYTHON_ASSUME_SAFE_MACROS && CYTHON_ASSUME_SAFE_SIZE)
+        if (PyBytes_AsStringAndSize(arg, &sub_ptr, &sub_len) == -1) return -1;
+        #else
         sub_ptr = PyBytes_AS_STRING(arg);
         sub_len = PyBytes_GET_SIZE(arg);
-        #if !CYTHON_ASSUME_SAFE_SIZE
-        if (unlikely(sub_len < 0)) return -1;
         #endif
+    } 
+    #if CYTHON_COMPILING_IN_LIMITED_API && __PYX_LIMITED_VERSION_HEX < 0x030B0000
+    else if (PyByteArray_Check(arg)) {
+        // The Limited API fallback is inefficient,
+        // so special-case bytearray to be a bit faster. Keep this to the Limited
+        // API only since the buffer protocol code is good enough otherwise.
+        sub_ptr = PyByteArray_AsString(arg);
+        if (unlikely(!sub_ptr)) return -1;
+        sub_len = PyByteArray_Size(arg);
+        if (unlikely(sub_len < 0)) return -1;
     } else {
+        // Where buffer protocol is unavailable, just convert to bytes
+        // (which is probably inefficient, but does work)
+        // First check that the object is a buffer (since PyBytes_FromObject)
+        // is more flexible than what endswith accepts.
+        PyObject *as_memoryview = PyMemoryView_FromObject(arg);
+        if (!as_memoryview) return -1;
+        Py_DECREF(as_memoryview);
```

Comment:
I've added a further check here (which makes it even less inefficient), but `PyBytes_FromObject` accepts a lot more types that `bytes.endswith` so it's necessary to get the same behaviour. Hopefully this path with be almost never used.

Reply:
We already handled tuples, bytes and bytearray. I'd also expect those to be the most common argument types here. Whatever else supports the buffer protocol can probably live with the performance penalty.